### PR TITLE
Autorotate

### PR DESCRIPTION
**Pull request by guiccbr**
Fix bugs in the implementation of rotate() and brake(). Also implement the auto-rotation feature. Please check the commit messages for more information.

To test the changes that have been made on this PR, a special gzsitl test-branch has been created.
https://github.com/guiccbr/gazebo-sitl/tree/test-autorotate

### CODE REVIEW

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 1]
```diff
@@ -54,7 +54,7 @@ const uint16_t remote_max_response_time_ms = 10000;
 const double waypoint_acceptance_radius_m = 0.01;
 const double detour_arrival_max_dist_m = 0.5;
 const double rotation_arrival_max_dif_deg = 10;
-const double is_stopped_max_speed_mps = 0.1;
+const double is_stopped_max_speed_mps = 0.3;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc]: This is not really related to improving output clarity.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 2]
```diff
@@ -296,19 +293,22 @@ void msghandler::handle(mav_vehicle &mav, const mavlink_message_t *msg)
...
         if (ack.target_system == mav.system_id && mav.is_sending_mission()) {
             mav.set_mode(mode::AUTO, 0);
             mav.sending_mission = false;
+            print_verbose("Mission started\n");
         }
+
+        // A MISSION_ACK has been broadcast what means that the mission might
+        // have changed, and so, our currently stored mission item might be
+        // outdated.
+        mav.curr_mission_item_outdated = true;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc]: same as above

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 3]
```diff
@@ -857,25 +862,36 @@ void mav_vehicle::brake(bool autocontinue)
 
     this->autocontinue_after_brake = autocontinue;
 
+    // Check what we are detouring, in order to the current action saved to
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc]: check "if" ?

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 4]
```diff
@@ -758,6 +758,17 @@ void mav_vehicle::rotate(double angle_deg, bool autocontinue)
     // Update timestamp
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc:801]: This commit can be squashed in the previous one, just add the explanation that you are changing for brakes and rotation. It's easier to understand the changes if they are together.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 5]
```diff
@@ -170,6 +171,30 @@ double dist(global_pos_int p1, global_pos_int p2)
...
+    double wp_rel_angle = rad2deg(ref_att.yaw) - wp_angle;
+
+    if (wp_rel_angle > 180.0) {
+        wp_rel_angle = wp_rel_angle - 360.0;
+    } else if (wp_rel_angle < -180.0) {
+        wp_rel_angle = wp_rel_angle + 360.0;
+    }
+
+    return wp_rel_angle;
+}
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc:191]: Any reason why you work with degrees instead of rads in here? it seems working with rads would avoid those conversions.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 6]
```diff
@@ -1161,6 +1186,30 @@ void mav_vehicle::update()
...
         }
     }
+
+    // Check if autorotation is active in order to request rotation if needed,
+    // if not already rotating. Need also to make sure that a mission waypoint
+    // is not currently being sent.
+    if (this->waypoint_autorotate &&
+        (this->mstatus == mission_status::NORMAL ||
+         this->mstatus == mission_status::DETOURING) &&
+        !this->sending_mission) {
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc:1244]: You could go with early return here, but it's also fine to keep it like this (in case you wanna add more code to this function later)

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 7]
```diff
@@ -739,6 +745,9 @@ void mav_vehicle::takeoff()
 
 void mav_vehicle::rotate(double angle_deg, bool autocontinue)
 {
+    // We have taken control
+    this->is_our_control = true;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc:767]: It seems a bit weird that you set control to true here. It you are asking the drone to rotate, you should check if you have the control of the drone, not set it to true. Can't we identify that we got the control back from any of the mavlink messages we receive? 
Same applies for the other functions below.

File: mavlink_vehicles/mavlink_vehicles.hh [hunk 8]
```diff
@@ -81,6 +81,7 @@ enum class mission_status {
     BRAKING,
     DETOURING,
     ROTATING,
+    MISSION,
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.hh]: this looks unrelated to this patch.

File: mavlink_vehicles/mavlink_vehicles.hh [hunk 9]
```diff
@@ -146,6 +147,7 @@ class mav_vehicle
                                bool autorotate = false);
     void send_mission_waypoint(global_pos_int global, bool autorotate = false);
     bool is_sending_mission() const;
+    bool is_mission_active() const;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.hh]: same as above.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 10]
```diff
@@ -1238,6 +1237,28 @@ void mav_vehicle::update()
...
+        if(this->time_stopped_ms == -1) {
+            this->stop_time = now;
+        }
+
+        this->time_stopped_ms =
+            duration_cast<milliseconds>(now - this->stop_time).count();
+
+    } else {
+        // Reset stop time counter
+        this->time_stopped_ms = -1;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc]: shouldn't you clear stop_time as well?

File: mavlink_vehicles/mavlink_vehicles.hh [hunk 11]
```diff
@@ -187,6 +187,11 @@ class mav_vehicle
     bool autocontinue_after_rotation = false;
     mission_status autocontinue_action = mission_status::NORMAL;
 
+    std::chrono::time_point<std::chrono::system_clock> stop_time =
+        std::chrono::system_clock::from_time_t(0);
+    int time_stopped_ms = -1;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.hh]: the naming on those 2 variables is very confusing.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 12]
```diff
@@ -1167,6 +1173,7 @@ void mav_vehicle::update()
 
         // Get back to AUTO mode to continue the mission
         if (this->detour_waypoint_autocontinue) {
+            this->is_our_control = true;
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc:1189]: this doesn't seems related to this patch.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 13]
```diff
@@ -468,8 +468,10 @@ bool mav_vehicle::is_ready()
         return false;
     }
 
-    // Also check if mission waypoint has been received
+    // Also check if mission waypoint has been received. If it has not, we need
+    // to request it after takeoff.
     if (!get_mission_waypoint().is_initialized()) {
+        request_mission_item(this->mission_waypoint_id);
```
Comment (rchiossi) [inline mavlink_vehicles/mavlink_vehicles.cc]: I don't think this belong in a is_ready() function. This function should have no functionality other than checking for a set of conditions, this is beyond that scope.

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 14]
```diff
@@ -65,7 +67,7 @@ const uint16_t arm_disarm = 1000;
 const uint16_t heartbeat = 1000;
 const uint16_t set_mode = 1000;
 const uint16_t takeoff = 1000;
-const uint16_t rotate = 1000;
+const uint16_t rotate = 200;
```
Comment (dakerfp) [inline mavlink_vehicles/mavlink_vehicles.cc:70]: In which unit does this value belong?

File: mavlink_vehicles/mavlink_vehicles.cc [hunk 15]
```diff
@@ -719,8 +758,18 @@ void mav_vehicle::takeoff()
                   defaults::takeoff_init_alt_m, request_intervals_ms::takeoff);
 }
 
-void mav_vehicle::rotate(double angle_deg)
+void mav_vehicle::set_autorotate(bool mission, bool detour)
```
Comment (dakerfp) [inline mavlink_vehicles/mavlink_vehicles.cc]: There should be two distinct setters.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 1 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: Since this value has been changed in a more recent commit, I've just removed it from this one.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 2 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: I've given a separate commit to it.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 4 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:801]: Not sure what you mean.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 3 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: Sorry, that comment was a complete mess.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 5 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:191]: The return value of this function will be used a parameter for a subsequent call to rotate(double angle_deg) in [here](https://github.com/01org/mavlink-vehicles/pull/5/files#diff-2801e8d06748835400032a671da7d476R1257). Since rotate() receives and angle in degrees, I decided to have get_waypoint_rel_angle to return in degrees as well. However, the decision to use degrees in rotate() was completely arbitrary. Rethinking about it, I agree that using radians instead, both in rotate() and get_waypoint_rel_angle(), would avoid some unnecessary symmetric conversions since the mavlink protocol itself uses radians in most of its messages.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 15 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: agree. fixed in 2f6cbc5917e12ce1268256f8bfe9c2b9841fbb0b

File: mavlink_vehicles/mavlink_vehicles.hh (hunk 8 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.hh]: Good catch. This shouldn't be here. It's not being used.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 12 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:1189]: I agree. But I'm afraid I won't be able to move this commit to another PR without a lot of rebase. Do you think it is really necessary?

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 14 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:70]: Milliseconds, as the namespace says: request_intervals_ms::
Is that what you mean?

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 10 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: Hmn, this whole block looks very confusing to me as well - though I myself have written it. I will take a closer look into it asap.

File: mavlink_vehicles/mavlink_vehicles.hh (hunk 11 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.hh]: As said in [here](https://github.com/01org/mavlink-vehicles/pull/5#discussion-diff-78393869R1260)

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 6 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:1244]: Indeed, we might need to add more code to this function later. I don't like those huge boolean expressions either, but I couldn't find a way of doing it neatly. Anyway, I'll take another look into it.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 7 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:767]: The drone is not exactly aware of whom is controlling it. It can be controlled by anyone that is connected to it. What we are doing here is only relevant to the mavlink_vehicles layer - to prevent multiple instances to try to autorotate/autocontinue during a detour or mission.

Every time a instance of a mavlink_vehicle sends a mission command to the vehicle, arducopter broadcasts a mission ACK saying that it has succesfuly received the command. This mission ACK is then parsed by every mavlink_vehicle instance connected to arducopter. If that ack is not being expected - i.e., the instance has not sent any mission command whose ack has not been received yet - that instance knows that someone else is trying to control the vehicle, so it stops to try to autorotate/autocontinue until either one of its command functions - rotate(), brake(), send_waypoint(), ... - is called.

In other words, the responsibility to autorotate/autocontinue is always in the hands of the last mavlink_vehicle instance that has sent a mission command.

File: mavlink_vehicles/mavlink_vehicles.hh (hunk 9 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.hh]: same as [above](https://github.com/01org/mavlink-vehicles/pull/5#discussion-diff-78384941R84).

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 4 above)
Comment (rchiossi) [reply mavlink_vehicles/mavlink_vehicles.cc:801]: No idea how this comment ended up here. It was about another commit.

Comment (rchiossi) [reply mavlink_vehicles/mavlink_vehicles.cc:801]: comments are all over the place. I mean merge "Continue to detour or mission waypoint after brake" with "Continue to detour or mission waypoint after rotating"

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 7 above)
Comment (rchiossi) [reply mavlink_vehicles/mavlink_vehicles.cc:767]: Ok, I got the idea. Still looks weird, but I think we can live with that for now.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 12 above)
Comment (rchiossi) [reply mavlink_vehicles/mavlink_vehicles.cc:1189]: Does it have any conflict if you use 'git rebase -e' to move this line to "Detect when someone else has taken control"

Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:1189]: No conflict. Force-pushed!

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 4 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:801]: Got your point. Force-pushed.

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 5 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc:191]: Fix: e43b3d428744d15d10c03a8a7d81da013c06fac3

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 10 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: Closer look taken. Fix submitted as a force push: f6bd7774400ba7e2decbb27697f4746a387ec51b

File: mavlink_vehicles/mavlink_vehicles.hh (hunk 11 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.hh]: Fixed in f6bd7774400ba7e2decbb27697f4746a387ec51b

File: mavlink_vehicles/mavlink_vehicles.cc (hunk 13 above)
Comment (guiccbr) [reply mavlink_vehicles/mavlink_vehicles.cc]: Fixed. 0e60676436ae4b5e6e2c1571bf0755d1805ee866

**Comment by rchiossi** [comment]
LGTM
